server/gitrest/packages/gitrest-base/src/utils/redisFs/redisFsManager.ts

```diff
@@ -0,0 +1,457 @@
+			| Stream,
+		options?:
+			| (ObjectEncodingOptions & {
+					mode?: Mode | undefined;
+					flag?: OpenMode | undefined;
+			  } & Abortable)
+			| BufferEncoding
+			| null,
+	): Promise<void> {
+		const filepathString = filepath.toString();
+		// Do not write packed-ref files which are not supported in r11s scenarios
+		if (filepathString.includes(packedRefsFileName)) {
+			return;
+		}
+
+		const result = await executeRedisFsApi(
+			async () => this.redisFs.set(filepath.toString(), data),
+			RedisFsApis.WriteFile,
+			RedisFSConstants.RedisFsApi,
+			this.redisFsConfig.enableRedisFsMetrics,
+			this.redisFsConfig.redisApiMetricsSamplingPeriod,
+			{
+				filepathString,
+			},
+		);
+
+		return result;
+	}
+
+	/**
+	 * Removes a link or a file. If path refers to a symbolic link, then the link is removed without affecting the file
+	 * or directory to which that link refers. If the path refers to a file path that is not a symbolic link, the file
+	 * is deleted. GitRest does not use symlinks in its Git
+	 * repos. Therefore, unlink always removes the target file.
+	 * For more info on how isomorphic-git uses this, see:
+	 * https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L143
+	 */
+	public async unlink(filepath: PathLike): Promise<void> {
+		const filepathString = filepath.toString();
+
+		await executeRedisFsApi(
+			async () => this.redisFs.delete(filepath.toString()),
+			RedisFsApis.Unlink,
+			RedisFSConstants.RedisFsApi,
+			this.redisFsConfig.enableRedisFsMetrics,
+			this.redisFsConfig.redisApiMetricsSamplingPeriod,
+			{
+				filepathString,
+			},
+		);
+	}
+
+	/**
+	 * Reads the contents of a directory.
+	 * For more info on how isomorphic-git uses this, see:
+	 * https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L167
+	 * https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L186
+	 * isomorphic-git never provides options, and always expects string[] results.
+	 */
+	public async readdir(
+		folderpath: PathLike,
+		options?:
+			| (ObjectEncodingOptions & { withFileTypes?: false | undefined })
+			| BufferEncoding
+			| null,
+	): Promise<string[]>;
+	public async readdir(
+		folderpath: PathLike,
+		options: { encoding: "buffer"; withFileTypes?: false | undefined } | "buffer",
+	): Promise<Buffer[]>;
+	public async readdir(
+		folderpath: PathLike,
+		options?:
+			| (ObjectEncodingOptions & { withFileTypes?: false | undefined })
+			| BufferEncoding
+			| null,
+	): Promise<string[] | Buffer[]>;
+	public async readdir(
+		folderpath: PathLike,
+		options: ObjectEncodingOptions & { withFileTypes: true },
+	): Promise<Dirent[]>;
+	public async readdir(
+		folderpath: PathLike,
+		options?: any,
+	): Promise<string[] | Buffer[] | Dirent[]> {
+		const folderpathString = folderpath.toString();
+
+		const result = await executeRedisFsApi(
+			async () => this.redisFs.keys_by_prefix(folderpath.toString()),
```

Comment:
Nice 😄 I didn't know about this `keys()` feature/command in Redis.

One question: if we have a file system structure like:

folder A
___folderB
______file1
___folderC
______file2
______file3

And we call `keys_by_prefix` with `folderA` as parameter, would it return `folderB` and `folderC` only, or would it return all the end files/full qualified path like `folderA/folderB/file1`, `folderA/folderC/file2`, `folderA/folderC/file3`? I think AzureBlobFs behaves like the latter, but it might be worth double checking

Reply:
It will return the full path

![image](https://github.com/microsoft/FluidFramework/assets/448106/3168bf38-b921-4b37-b2df-cbd7a64c59c1)